pkg/lwip/contrib/sock/lwip_sock.c

```diff
@@ -0,0 +1,526 @@
+            if (res < 0) {
+                netconn_delete(*conn);
+                return res;
+            }
+        }
+        if (remote != NULL) {
+            switch (netconn_connect(*conn, &remote_addr, remote_port)) {
+                case ERR_USE:
+                    res = -EADDRINUSE;
+                    break;
+                case ERR_VAL:
+                    res = -EINVAL;
+                    break;
+                default:
+                    break;
+            }
+            if (res < 0) {
+                netconn_delete(*conn);
+                return res;
+            }
+        }
+    }
+    return 0;
+}
+
+uint16_t lwip_sock_bind_addr_to_netif(const ip_addr_t *bind_addr)
+{
+    assert(bind_addr != NULL);
+
+    if (!ip_addr_isany(bind_addr)) {
+        for (struct netif *netif = netif_list; netif != NULL; netif = netif->next) {
+            if (IP_IS_V6(bind_addr)) {  /* XXX crappy API yields crappy code */
+#if LWIP_IPV6
+                if (netif_get_ip6_addr_match(netif, ip_2_ip6(bind_addr)) >= 0) {
+                    return (int)netif->num + 1;
+                }
+#endif
+            }
+            else {
+#if LWIP_IPV4
+                if (netif_ip4_addr(netif)->addr == ip_2_ip4(bind_addr)->addr) {
+                    return (int)netif->num + 1;
+                }
+#endif
+            }
+        }
+    }
+    return SOCK_ADDR_ANY_NETIF;
+}
+
+int lwip_sock_get_addr(struct netconn *conn, struct _sock_tl_ep *ep, u8_t local)
+{
+    ip_addr_t addr;
+    int res;
+#ifdef MODULE_LWIP_SOCK_IP
+    u16_t port = UINT16_MAX;
+    u16_t *port_ptr = &port;
+    /* addr needs to be NULL because netconn_getaddr returns error on connected
+     * conns as "as connecting is only a helper for upper layers [sic]" */
+    memset(&addr, 0, sizeof(addr));
+#else
+    u16_t *port_ptr = &ep->port;
+#endif
+
+    assert(ep != NULL);
+    if (conn == NULL) {
+        return 1;
+    }
+#ifdef MODULE_LWIP_SOCK_IP
+    if (!(conn->type & NETCONN_RAW)) {
+        port_ptr = &ep->port;
+    }
+#endif
+    if ((res = netconn_getaddr(conn, &addr, port_ptr, local)) != ERR_OK
+#ifdef MODULE_LWIP_SOCK_IP
+        /* XXX lwIP's API is very inconsistent here so we need to check if addr
+         * was changed */
+            && !local && ip_addr_isany_val(addr)
+#endif
+        ) {
+        return res;
+    }
+#if LWIP_IPV6 && LWIP_IPV4
+    ep->family = (addr.type == IPADDR_TYPE_V6) ? AF_INET6 : AF_INET;
+#elif LWIP_IPV6
+    ep->family = (conn->type & NETCONN_TYPE_IPV6) ? AF_INET6 : AF_INET;
+#elif LWIP_IPV4
+    ep->family = AF_INET;
+#endif
+    if (local) {
+        ep->netif = lwip_sock_bind_addr_to_netif(&addr);
+    }
+    else {
+        ep->netif = SOCK_ADDR_ANY_NETIF;
+    }
+    memcpy(&ep->addr, &addr, sizeof(ep->addr));
+    return 0;
+}
+
+int lwip_sock_recv(struct netconn *conn, uint32_t timeout, struct netbuf **buf)
+{
+    int res;
+
+    if (conn == NULL) {
+        return -EADDRNOTAVAIL;
+    }
+#if LWIP_SO_RCVTIMEO
```

Comment:
What's the reason to make this optional?
It's kind of surprising that LWIP will by default ignore the timeout in e.g. `sock_udp_recv()`